### PR TITLE
MGMT-17064: add spec manager

### PR DESCRIPTION
**Pull request by hexfusion**
This PR adds the spec manager which controls the management of current and desired spec.

### expected flow boostrap

- current: during `bootstrap` spec manager will ensure current configuration by first checking for the existence of the `spec.CurrentFile` if that exists we serialize that into `v1alpha1.RenderedDeviceSpec`. Otherwise we write an empty `v1alpha1.RenderedDeviceSpec` to disk as we have no other information.

- desired: during `bootstrap` spec manager will ensure current configuration by first checking if `spec.DesiredFile` exists otherwise we call the management API and ask for the latest spec. Once that is received we persist that to disk.

### expected flow runtime
- current: during runtime we will always get current `v1alpha1.RenderedDeviceSpec`  from disk. if that file is missing or corrupted this is an error case that will need more thought.

- desired: desired will be retrieved from the management API if there is no change returned (204) we `skipSync` otherwise the desired `v1alpha1.RenderedDeviceSpec` is populated from the API. If the agent is unable to connect to the management API after exceeding the max retries we will fallback to the latest copy on disk.

- reconciled: Once the controllers have reconciled the spec is the desired spec `updated` on disk as `current` spec.

```mermaid
graph TD
    A(Start Bootstrap) --> B{Check spec.CurrentFile}
    B -->|Exists| C(Serialize to v1alpha1.RenderedDeviceSpec)
    B -->|Not Exists| D(Write empty v1alpha1.RenderedDeviceSpec to disk)
    E(Start Desired Bootstrap) --> F{Check spec.DesiredFile}
    F -->|Not Exists| G(Call Management API)
    G --> H(Persist API spec to disk)
    
    I(Start Runtime) --> J{Get v1alpha1.RenderedDeviceSpec from disk}
    J -->|Missing/Corrupted| K(Error, needs thought)
    L(Start Desired Runtime) --> M(Retrieve spec from Management API)
    M -->|No Change 204| N(skipSync)
    M -->|Change| O(Populate v1alpha1.RenderedDeviceSpec from API)
    M -->|Max Retries Exceeded| P(Fallback to latest disk copy)
    
    C --> Q{Reconciled}
    H --> Q
    O --> Q
    P --> Q
    Q --> R(Update desired spec as current on disk)
```

## ChangeLog

- add fileio pkg to manage all reads and writes to the files system
- add spec manager (integration with agent will be a followup)

### CODE REVIEW

File: internal/agent/device/fileio/writer.go [hunk 1]
```diff
@@ -66,6 +67,7 @@ func (w *Writer) WriteIgnitionFiles(files ...ign3types.File) error {
 		// if err := createOrigFile(file.Path, file.Path); err != nil {
 		// 	return err
 		// }
+		klog.Info("Writing file: ", w.rootDir+file.Path)
```
Comment (hexfusion) [inline internal/agent/device/fileio/writer.go:70]: drop

File: internal/agent/device/fileio/writer.go [hunk 2]
```diff
@@ -75,7 +77,8 @@ func (w *Writer) WriteIgnitionFiles(files ...ign3types.File) error {
 
 // WriteFile writes the provided data to the file at the path with the provided permissions
 func (w *Writer) WriteFile(name string, data []byte, perm fs.FileMode) error {
-	return os.WriteFile(filepath.Join(w.rootDir, name), data, perm)
+	// convert to ign file so we can use the atomic writer we can do this more directly in future
```
Comment (hexfusion) [inline internal/agent/device/fileio/writer.go:80]: add TODO

File: internal/agent/device/spec/spec.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+					log.Warningf("%s received 204 from management API", logPrefix)
+				}
+
+				return true, nil
+			})
+			if err != nil {
+				log.Errorf("%sFailed to get desired spec after retry: %v", logPrefix, err)
+				// this could mean the management API is not available, or the internet connection is down.
+				// we really can't make progress here without the desired rendered spec. this should fail the service
+				// and require a restart.
```
Comment (hexfusion) [inline internal/agent/device/spec/spec.go:239]: add TODO: and track issue for making progress if state is corrupted.

File: internal/client/management.go [hunk 4]
```diff
@@ -81,3 +81,21 @@ func (m *Management) UpdateDeviceStatus(ctx context.Context, name string, buf *b
...
+
+// GetRenderedDeviceSpec returns the rendered device spec for the given device and the response code.
+func (m *Management) GetRenderedDeviceSpec(ctx context.Context, name string, params *v1alpha1.GetRenderedDeviceSpecParams, rcb ...client.RequestEditorFn) (*v1alpha1.RenderedDeviceSpec, int, error) {
+	start := time.Now()
+	resp, err := m.client.GetRenderedDeviceSpecWithResponse(ctx, name, params, rcb...)
+	if m.rpcMetricsCallbackFunc != nil {
+		m.rpcMetricsCallbackFunc("get_rendered_device_spec_duration", time.Since(start).Seconds(), err)
+	}
+	if err != nil {
+		return nil, 0, err
```
Comment (hexfusion) [inline internal/client/management.go:93]: 500 ?

File: internal/client/management.go [hunk 5]
```diff
@@ -81,3 +81,21 @@ func (m *Management) UpdateDeviceStatus(ctx context.Context, name string, buf *b
...
+	}
+
+	if resp.JSON200 != nil {
+		return resp.JSON200, resp.StatusCode(), nil
+	}
+
+	return nil, resp.StatusCode(), nil
```
Comment (hexfusion) [inline internal/client/management.go:100]: add docs on 204 handling

**Comment by avishayt** [comment]
LGTM.  Obviously some parts are missing, but so far so good.

**Comment by mangelajo** [comment]
I agree with @avishayt , the approach makes sense so far. It looks good.

**Review by mangelajo — APPROVED** [review]
I am good with merge-and-iterate here.

**Pull request merged**